### PR TITLE
Add Merkle tree implementation in Go

### LINKED ISSUE #395: Implement the Merkle proof algorithm
## 📚 Context
The Merkle tree is the data structure that generates proofs for the outputs. It currently has three different implementations: 
- [Solidity](https://github.com/cartesi/rollups-contracts/blob/main/contracts/library/LibMerkle32.sol): used by the rollups contracts
- [C](https://github.com/cartesi/machine-emulator-tools/blob/main/sys-utils/libcmt/src/merkle.c): used inside the Cartesi Machine
- [Rust](https://github.com/cartesi/rollups-node/blob/main/offchain/host-runner/src/merkle_tree/complete.rs): used by the Node v1.x

The original plan for Node v2 was to port the Rust implementation to Go. This was later abandoned as the Rust implementation was deeply intertwined with the old way of generating proofs (dividing the original keccak256 hash into four 64-bit pieces and creating a Merkle tree with them), overcomplicating it.

## ✔️ Solution
The Node's use case for Merkle proofs is straightforward: it only needs to generate proofs at an epoch's end. By then, we already have all the outputs available and can calculate the root hash and the sibling arrays. This only needs to be done once, as there won't be any new outputs for that epoch that would require recalculating it.

Therefore, we can design a more efficient algorithm than a complete, generic Merkle tree data structure that wouldn't fit our current needs as well. The algorithm in question is very similar to the one currently implemented in Solidity.

## 📈 Subtasks
- [x] Write the algorithm in golang
- [x] Test it against the Solidity implementation
- [x] Test it against the C implementation

### PR DESCRIPTION
**Pull request by torives**
Closes #395

### CODE REVIEW

File: internal/merkle/tree.go
Comment (diegonehab) [inline internal/merkle/tree.go:174]: 2^64 does not fit in a uint64. So your max_tree_height has to be 63, right?

File: internal/merkle/tree.go
Comment (diegonehab) [inline internal/merkle/tree.go:16]: See comment on `maxLeaves` below.

File: internal/merkle/tree.go
Comment (diegonehab) [inline internal/merkle/tree.go:119]: Does this work when the tree is empty?

File: internal/merkle/tree.go
Comment (diegonehab) [inline internal/merkle/tree.go:89]: This is *very* inefficient. When you append a new leaf, you only need to update one node per level, the "rightmost" node. Here, you are recomputing every internal node.

File: internal/merkle/tree.go
Comment (diegonehab) [inline internal/merkle/tree.go:47]: How is this root "pristine"? Shouldn't you copy it from the pristine tree?

File: internal/merkle/tree.go
Comment (diegonehab) [inline internal/merkle/tree.go:156]: Shouldn't you be dividing index by 2 at each iteration here?

Comment (diegonehab) [reply internal/merkle/tree.go:156]: And if so, how is it that this is passing the tests? Shouldn't the tests have caught this?

**Review by diegonehab — CHANGES_REQUESTED** [review]
I think that when designing a datastructure like this, you have to keep in mind the use cases in the node.

I don't see a situation in which the node will have to keep this tree up-to-date as each leaf is being inserted. If there was such a use case, then please see comments in the code as to how to make this much more efficient. O(N D) instead, where N is the number of leaves and D the tree depth, instead of the current O(N^2) algorithm.

The use case I can see is a "batch" operation. The node will simply append the leaves as they come, and once the epoch is closed, it will need to produce everything. This can also be done in O(N D) without there even being a tree structure.

Let me know what use cases you had in mind.

**Comment by guidanoli** [comment]
I think I agree with Diego on this one. You only need to generate the epoch hash at the end of every epoch, and this tree is brand new every epoch. This will, of course, be different after Output Unification v2, with an ever-growing output tree.

For now, I think this code can be a lot simpler. You might not even need to define a type for pristine trees. You can just calculate the node of the pristine tree at the current level on-the-fly when going from bottom to top.

**Comment by torives** [comment]
Thanks, @diegonehab and @guidanoli for your review. I've removed the tree data structure and went with the batch strategy previously mentioned. I've also re-written the tests (inspired by the Solidity test code) and am more confident it is doing what it should. Please, take another look.

File: internal/merkle/proof.go
Comment (guidanoli) [inline internal/merkle/proof.go:65]: Why not pre-allocate the whole matrix in `CreateProofs`?

File: internal/merkle/proof.go
Comment (guidanoli) [inline internal/merkle/proof.go:96]: Another use of the `at` function.

```suggestion
		rightChild := at(currentLevel, 2*idx+1, pristineNode)
```

File: internal/merkle/proof.go
Comment (guidanoli) [inline internal/merkle/proof.go:87]: When I think of a full level, I think it has $2^h$ non-pristine nodes. But it may be not full, but a pristine node is still used, if it has an odd number of non-pristine nodes.

File: internal/merkle/proof.go
Comment (guidanoli) [inline internal/merkle/proof.go:53]: I would also rephrase this to something like...

> If the current level has an odd number of nodes, a pristine node will be used.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:47]: The `at` function is quite elegant. I liked it very much when reading the Solidity code but forgot to implement it here as well 😅. Thanks.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:32]: It was a workaround as Golang requires both operands to be the same type. For brevity, I preferred `height` cast to int instead of `levelIdx` to uint, but converting `calculateSiblings` parameters to unsigned ints is better 👍🏽

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:65]: Because I would have to loop over the leaves one more time just for this purpose

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:87]: You're correct, I mixed up the concepts here.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:73]: I guess it could be if I were iterating over the nodes at the current level. But I'm always iterating over the leaves, as I found it easier to handle the matrix indexes.

File: internal/merkle/proof.go
Comment (guidanoli) [reply internal/merkle/proof.go:62]: That's true, but it would remove this `if` clause from your code.
On `CreateProofs`, you'd know it is `nil`.
On `CreateSiblings`, you'd like it is not `nil`.

File: internal/merkle/proof.go
Comment (guidanoli) [reply internal/merkle/proof.go]: I didn't know Golang didn't coerce 0 into `uint`. You could do this then...

```go
	for levelIdx := uint(0); levelIdx < height; levelIdx++ {
```

File: internal/merkle/proof.go
Comment (diegonehab) [reply internal/merkle/proof.go:62]: Why not use a single allocation? I.e., allocate an array with `len(leaves)*height` entries, and then access it with index (i,j) using i*height+j, where i is the leaf j is the level in the tree. Or you can transpose it, if you want, and access it as j*len(leaves)+i. Doesn't matter if you use the so-called column-major or row-major order. Either way it's a single allocation. It's how people represent matrices in "real life".

File: internal/merkle/proof.go
Comment (diegonehab) [inline internal/merkle/proof.go:32]: Come on, you are using a language that has multi-value assignments. Embrace it and write 

```golang
currentLevel, nextLevel = nextLevel, currentLevel[:(len(currentLevel)+1)/2]
```

File: internal/merkle/proof.go
Comment (diegonehab) [inline internal/merkle/proof.go:89]: Add documentation?

File: internal/merkle/proof.go
Comment (diegonehab) [inline internal/merkle/proof.go:80]: I think this is a matter of taste, but I see an error prone pattern here. The input is `currentLevel`. `nextLevel` is the output. And yet, you are reading the size of the *output* to figure out how many times to loop. I think it would be best to reslice `nextLevel` inside this function, rather than on the outside. Right?

Comment (diegonehab) [reply internal/merkle/proof.go:80]: Is there any convention in Go regarding where inputs and outputs go in the order of arguments in a function? In C++, we typically do inputs-first-then-outputs.

File: internal/merkle/proof.go
Comment (diegonehab) [inline internal/merkle/proof.go:83]: I *think* both these definitions create copies, right? Shouldn't you use `[:]` to avoid that? Or am I mistaken?

Comment (diegonehab) [reply internal/merkle/proof.go:83]: I see this pattern elsewhere in the code. Or is it copy-on-write? Do you know?

File: internal/merkle/proof.go
Comment (diegonehab) [reply internal/merkle/proof.go]: I don't get it either. Can you explain? This doesn't seem right at all when level is `> 1`...

**Review by diegonehab — DISMISSED** [review]
Please see comments. I am worried that, once again, your tests are not comprehensive enough. Unless the potential bug @guidanoli and I identified by code inspection are really not bugs.

**Comment by guidanoli** [comment]
> your tests are not comprehensive enough.

I agree. One nice approach is to do random tests. Pick a random $n \in [1,N]$ for a generous constant $N$ (say, 100). Then, pick $n$ random hashes $h_i$ for $i \in [1,n]$. Then, construct a Merkle tree with these $n$ hashes as leaves. Construct proofs for each of them. Then, verify each proof. Also, you can pick a random sibling in the proof and replace it by a different hash, and check that the proof doesn't match, just to catch false positives in the verification algorithm.

File: internal/merkle/proof.go
Comment (guidanoli) [reply internal/merkle/proof.go:80]: You could even write the next level in-place, given that you've already extracted the siblings at the current level with `calculateSiblings`. Therefore, this function could have the following signature:

```go
func calculateParents(level []model.Hash, pristineNode model.Hash)
```

I don't know if it is possible to slice an array in-place in Golang. If not, this function could return the size of the slice, so that the caller does the slicing. With this, the `CreateProofs` function wouldn't even need to maintain a `nextLevel`, just the `currentLevel`.

File: internal/merkle/proof.go
Comment (guidanoli) [reply internal/merkle/proof.go:83]: That's true. It seems that the `model.Hash` type is, in fact, just an array of 32 bytes. So, maybe `[:]` is necessary to avoid copying hashes around? I don't know much about Golang.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: You are correct, I re-checked my math and realized the error. I'll fix the code and enhance the test suite as suggested by @guidanoli.

File: internal/merkle/proof.go
Comment (diegonehab) [reply internal/merkle/proof.go:80]: I think using two levels is fine. Diminishing returns.

**Comment by diegonehab** [comment]
The thing is that you should try to cover the basic cases. You haven’t tested the sibling hashes by bubbling up from leaf to root to see if it matches the root hash. You should take a small height (say, 3 or 4) and test *everything*. All number of leaves. The siblings for all of them. Too many leaves. Zero leaves. Ideally, you should generate the expected results using some other code that we already have either in solidity, Lua, or C++.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:83]: > I _think_ both these definitions create copies, right?

Yes, they're both copies.

> Shouldn't you use `[:]` to avoid that? Or am I mistaken?

I could. But since I need to modify the `at` function to use pointers for the same reason (avoiding copies), I would rather use pointers directly.

Comment (diegonehab) [reply internal/merkle/proof.go:83]: Don't know much about Golang, but why pointers? I realize there is no concept of a const slice (go figure...).

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: Done at a764662892a930eda44dd48b1a1278f0da722e41

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: Done at a764662892a930eda44dd48b1a1278f0da722e41

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: I liked @guidanoli suggestion and ended up implementing it at 293e57080aa44d2aded47e53f6ef202c4c99295b.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: This code changed a bit because of [this](https://github.com/cartesi/rollups-node/pull/444#discussion_r1679388012). Changes can be found at 293e57080aa44d2aded47e53f6ef202c4c99295b.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: > I realize there is no concept of a const slice (go figure...).
AFAIK in Golang arrays are meant to be fixed and slices are meant to be dynamic. That is just the way it is 🤷🏽‍♂️ 

> Don't know much about Golang, but why pointers?
It is the only way we can avoid copies In the current situation (a slice of byte arrays). It didn't look so bad in the end. Done at a66d705dc0b807351ce69d39378a031e3df117e3.

File: internal/merkle/proof_test.go
Comment (guidanoli) [inline internal/merkle/proof_test.go:979]: I think we can use bit shifting here.

```suggestion
		leafCount = bound(leafCount, 1, 1 << height)
```

File: internal/merkle/proof_test.go
Comment (guidanoli) [inline internal/merkle/proof_test.go:1001]: You could select a random sibling as well.

File: internal/merkle/proof_test.go
Comment (guidanoli) [inline internal/merkle/proof_test.go:996]: I would check that the leaf is non-zero just to be safe.

**Review by guidanoli — COMMENTED** [review]
Looking good!

File: internal/merkle/proof.go
Comment (diegonehab) [reply internal/merkle/proof.go]: When you assign a reslice of an array to a variable, you don’t create a copy, unless I’m crazy.

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go:170]: So I can grab a leaf by its index later when verifying the proofs. Since `CreateProofs` uses `leaves`' underlying array to do its thing, I can't reuse it.

Comment (guidanoli) [reply internal/merkle/proof_test.go:170]: I see, so we leave it up to the caller to decide whether they want to copy the array of leaves before calling `CreateProofs`.

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go:979]: 🤯 🔫

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done at 35c518c6956338e4a67982cadf920ddd7942f128

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done at 6fc865150288bda76d2a6ebd46d3cc1b34935f8a

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done at 6fc865150288bda76d2a6ebd46d3cc1b34935f8a

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done at 26708fc3d1d198807daf41d1303e070ca5048a90

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done at 92c00d6ec1ebed0c913cffac657d70596da188d7.

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done at ad110882c5bdab743286d0c6dc25582b07587f32

**Review by guidanoli — DISMISSED** [review]
Will re-approve if needed after squash.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: My previous comment wasn't completely true. I could also re-slice it to avoid a copy as you originally proposed, but then I would end up with a `[]byte` slice, instead of a `model.Hash`. This would defeat the purpose of having a custom type to represent hashes so I went with a pointer to `model.Hash` instead.

Comment (diegonehab) [reply internal/merkle/proof.go]: To be fair, that is exactly what the implementation of `crypto.Keccak256Hash`. This seems to be idiomatic, no? I am not a Go specialist.

Comment (torives) [reply internal/merkle/proof.go]: I agree with you. Just because I needed to use pointers for the `at` function it doesn't mean I have to use them here also when there's a more idiomatic way. I changed it at 9eb81fc6e7b4c11d24a4649bf7823edaa1c6c05a.

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go:23]: > The given height could be checked against the number of leaves and return the error `too many leaves for height`. 

We could, but that would force me to define and check the maximum height, adding a second `if` statement and error. I'd rather keep it as it is, looking as close as possible to the implementation we have in the [contracts](https://github.com/cartesi/rollups-contracts/blob/main/contracts/library/LibMerkle32.sol#L19).

File: internal/merkle/proof.go
Comment (torives) [reply internal/merkle/proof.go]: Done at a422f86e1510ec873cad8979e3f9e8a7d4999c4f

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go:796]: I'm not sure I understood your goal. All fuzzy tests are executed along the unit tests, using the data added programatically or located in the `testdata` directory if that's what you meant.

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done in 09efe7b9a0e8a783ff101257114ee0c00f88d261

File: internal/merkle/proof_test.go
Comment (torives) [reply internal/merkle/proof_test.go]: Done in 09efe7b9a0e8a783ff101257114ee0c00f88d261.

File: .golangci.yml
Comment (torives) [reply .golangci.yml]: Actually, I don't this is needed anymore. I removed it in [1a3371e](https://github.com/cartesi/rollups-node/pull/444/commits/1a3371e52ebba4e925d8722401910e39dcad6c18).

**Review by fmoura — DISMISSED** [review]

**Review by marcelstanley — APPROVED** [review]

**Review by fmoura — DISMISSED** [review]

File: internal/merkle/proof.go
Comment (guidanoli) [inline internal/merkle/proof.go:32]: I would expand on this. Feel free to adapt it however you like.

```suggestion
			// Each pair of siblings shares the same parent.
			// For any given parent with index i,
			// the index of its left node is 2*i,
			// and the index of its right node is 2*i+1.
			// Since their indices differ only by the LSB,
			// by flipping the LSB with ^1,
			// we calculate the index of one another.
			// In order to get the index of a parent,
			// you remove the LSB with >>1.
			// For an ancestor of height difference h,
			// you remove the h LSBs with >>h.
```

**Review by marcelstanley — APPROVED** [review]

**Review by fmoura — APPROVED** [review]

**Pull request merged**
